logfire/_internal/auth.py

```diff
@@ -26,12 +37,133 @@ class UserTokenData(TypedDict):
+        else:
+            token_repr += self.token[:5]
+        token_repr += '****'
+        return token_repr
+
+
+@dataclass
+class UserTokenCollection:
+    """A collection of user tokens."""
+    user_tokens: list[UserToken]
+
+    @classmethod
+    def from_tokens(cls, tokens: TokensFileData) -> Self:
+        return cls(
+            user_tokens=[
+                UserToken.from_user_token_data(url, data)  # pyright: ignore[reportArgumentType], waiting for PEP 728
+                for url, data in tokens.items()
+            ]
+        )
+
+    @classmethod
+    def from_tokens_file(cls, file: Path) -> Self:
+        return cls.from_tokens(cast(TokensFileData, read_toml_file(file)))
+
+    def get_token(self, base_url: str | None = None) -> UserToken:
+        if base_url is not None:
+            token = next((t for t in self.user_tokens if t.base_url == base_url), None)
+            if token is None:
+                raise LogfireConfigError(
+                    f'No user token was found matching the {base_url} Logfire URL. '
+                    'Please run `logfire auth` to authenticate.'
+                )
+        else:
+            if len(self.user_tokens) == 1:
+                token = self.user_tokens[0]
+            elif len(self.user_tokens) >= 2:
+                choices_str = '\n'.join(
+                    f'{i}. {token} ({"expired" if token.is_expired else "valid"})'
+                    for i, token in enumerate(self.user_tokens, start=1)
+                )
+                int_choice = IntPrompt.ask(
+                    f'Multiple user tokens found. Please select one:\n{choices_str}\n',
+                    choices=[str(i) for i in range(1, len(self.user_tokens) + 1)],
+                )
+                token = self.user_tokens[int_choice - 1]
+            else:  # self.user_tokens == []
+                raise LogfireConfigError('No user tokens are available. Please run `logfire auth` to authenticate.')
+
+        if token.is_expired:
+            raise LogfireConfigError(f'User token {token} is expired. Pleas run `logfire auth` to authenticate.')
+        return token
+
+    def is_logged_in(self, base_url: str | None = None) -> bool:
+        if base_url is not None:
+            tokens = (t for t in self.user_tokens if t.base_url == base_url)
+        else:
+            tokens = self.user_tokens
+        return any(not t.is_expired for t in tokens)
+
+    def add_token(self, base_url: str, token: UserTokenData) -> UserToken:
+        existing_token = next((t for t in self.user_tokens if t.base_url == base_url), None)
+        if existing_token:
+            token_index = self.user_tokens.index(existing_token)
+            self.user_tokens.remove(existing_token)
+        else:
+            token_index = len(self.user_tokens)
+
+        user_token = UserToken.from_user_token_data(base_url, token)
+        self.user_tokens.insert(token_index, user_token)
```

Comment:
This isn't really ideal logic, perhaps I should have the internal `UserTokenCollection` structure matching the toml file (i.e. mapping of url->token)?